src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Refactoring/ExtractToCodeBehindCodeActionProvider.cs

```diff
@@ -0,0 +1,75 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Microsoft.CodeAnalysis;
+using Newtonsoft.Json.Linq;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Refactoring
+{
+    class ExtractToCodeBehindCodeActionProvider : RazorCodeActionProvider
+    {
+        override public Task<CommandOrCodeActionContainer> ProvideAsync(RazorCodeActionContext context, CancellationToken cancellationToken)
+        {
+            var directiveNode = (RazorDirectiveSyntax)context.Document.GetNodeAtLocation(context.Location, n => n.Kind == SyntaxKind.RazorDirective);
+            if (directiveNode is null)
+            {
+                return null;
+            }
+
+            if (directiveNode.DirectiveDescriptor.Directive != "code" && directiveNode.DirectiveDescriptor.Directive != "function")
+            {
+                return null;
+            }
+
+            var cSharpCodeBlockNode = directiveNode.Body.DescendantNodes().FirstOrDefault(n => n is CSharpCodeBlockSyntax);
```

Comment:
How does this behave for malformed directive nodes? aka `@code` without a body. I'm pretty sure it'll just work because we represent those as "malformed directive" nodes but something to verify it's accounted for

Reply:
Right now if you just have `@code`, it allows you to extract it and creates an empty partial class in the resultant file just as a side-effect of how it's implemented. How should I check if it's malformed explicitly?

Reply:
Ah, that's not awful. I'd leave as is then for now.